axlearn/common/checkpointer.py

```diff
@@ -416,11 +431,19 @@ def restore_from_dir(
             spec.tf_ckpt_map, dir=os.path.join(ckpt_dir, f"tf_{jax.process_index()}")
         )
 
+        # Override dtype to target cast dtype.
+        # From jax docs,
+        # Cast while reloading on process to avoid 2 copies on device if the
+        # casting is done on device.
+        spec_dtypes = spec.dtypes
+        if dtype is not None:
+            spec_dtypes = [dtype] * len(spec_dtypes)
```

Comment:
How would this handle integer vs float dtypes?

Reply:
how about we add a more complicated check? And only add this for bfloat16 and float16 to float32 weights. wdyt?

Reply:
How about taking an `Optional[Callable[[jnp.dtype], jnp.dtype]]` that maps one dtype to another?

Reply:
Sounds good. Let me try